File: src/test/java/com/github/kayjamlang/tests/expressions/access/AccessExpressionTest.java
package com.github.kayjamlang.tests.expressions.access;

import com.github.kayjamlang.core.expressions.AccessExpression;
import com.github.kayjamlang.core.expressions.Expression;
import com.github.kayjamlang.core.KayJamLexer;
import com.github.kayjamlang.core.KayJamParser;
import com.github.kayjamlang.core.expressions.CallOrCreateExpression;
import com.github.kayjamlang.core.expressions.VariableLinkExpression;
import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.*;

public class AccessExpressionTest {

    private static KayJamParser parser;

    @BeforeClass
    public static void prepare(){
        parser = new KayJamParser(new KayJamLexer("getRoot().child"));
    }

    @Test
    public void test() throws Exception {
        Expression expression = parser.readExpression();

        assertNotNull(expression);
        assertSame(AccessExpression.class, expression.getClass());

        AccessExpression accessExpression = (AccessExpression) expression;
        assertSame(CallOrCreateExpression.class, accessExpression.root.getClass());
        assertSame(VariableLinkExpression.class, accessExpression.child.getClass());

        CallOrCreateExpression root = (CallOrCreateExpression) accessExpression.root;
        assertEquals("getRoot", root.name);
        assertEquals(0, root.arguments.size());

        VariableLinkExpression child = (VariableLinkExpression) accessExpression.child;
        assertEquals("child", child.name);
    }
}

File: src/test/scala/com/github/kayjamlang/tests/expressions/access/AccessExpressionTest.scala
package com.github.kayjamlang.tests.expressions.access

import com.github.kayjamlang.core.{KayJamLexer, KayJamParser}
import com.github.kayjamlang.core.expressions.{AccessExpression, CallOrCreateExpression, VariableLinkExpression}
import org.junit.Assert._
import org.junit.{BeforeClass, Test}


object AccessExpressionTest {
  private var parser: KayJamParser = null

  @BeforeClass def prepare(): Unit = {
    parser = new KayJamParser(new KayJamLexer("getRoot().child"))
  }
}

class AccessExpressionTest {
  @Test
  @throws[Exception]
  def test(): Unit = {
    val expression = AccessExpressionTest.parser.readExpression
    assertNotNull(expression)
    assertSame(classOf[AccessExpression], expression.getClass)
    val accessExpression = expression.asInstanceOf[AccessExpression]
    assertSame(classOf[CallOrCreateExpression], accessExpression.root.getClass)
    assertSame(classOf[VariableLinkExpression], accessExpression.child.getClass)
    val root = accessExpression.root.asInstanceOf[CallOrCreateExpression]
    assertEquals("getRoot", root.name)
    assertEquals(0, root.arguments.size)
    val child = accessExpression.child.asInstanceOf[VariableLinkExpression]
    assertEquals("child", child.name)
  }
}